Lib/ufo2ft/featureWriters/kernFeatureWriter.py

```diff
@@ -32,62 +47,229 @@ def unicodeBidiType(uv):
         else:
             raise AssertionError(side1)
 
+        self.side2: ast.GlyphName | ast.GlyphClassName | ast.GlyphClass
         if isinstance(side2, str):
             self.side2 = ast.GlyphName(side2)
         elif isinstance(side2, ast.GlyphClassDefinition):
             self.side2 = ast.GlyphClassName(side2)
+        elif isinstance(side2, (ast.GlyphClass, ast.GlyphClassName)):
+            self.side2 = side2
+        elif isinstance(side2, (list, set)):
+            if len(side2) == 1:
+                self.side2 = ast.GlyphName(list(side2)[0])
+            else:
+                self.side2 = ast.GlyphClass([ast.GlyphName(g) for g in sorted(side2)])
         else:
             raise AssertionError(side2)
 
-        self.value = value
-        self.directions = directions or set()
-        self.bidiTypes = bidiTypes or set()
+        self.value: float = value
+        self.scripts: set[str] = scripts or set()
+        self.directions: set[str] = directions or set()
+        self.bidiTypes: set[str] = bidiTypes or set()
+
+    def __lt__(self, other: KerningPair) -> bool:
+        if not isinstance(other, KerningPair):
+            return NotImplemented
+
+        # NOTE: Comparing classes relies on their glyph names being sorted in
+        # __init__.
+        selfTuple = (
+            self.firstIsClass,
+            self.secondIsClass,
+            self.side1.glyph
+            if isinstance(self.side1, ast.GlyphName)
+            else self.side1.glyphSet()[0].glyph,
+            self.side2.glyph
+            if isinstance(self.side2, ast.GlyphName)
+            else self.side2.glyphSet()[0].glyph,
+        )
+        otherTuple = (
+            other.firstIsClass,
+            other.secondIsClass,
+            other.side1.glyph
+            if isinstance(other.side1, ast.GlyphName)
+            else other.side1.glyphSet()[0].glyph,
+            other.side2.glyph
+            if isinstance(other.side2, ast.GlyphName)
+            else other.side2.glyphSet()[0].glyph,
+        )
+        return selfTuple < otherTuple
+
+    def partitionByScript(
+        self, glyphScripts: dict[str, set[str]]
+    ) -> Iterator[tuple[str, KerningPair]]:
+        """Split a potentially mixed-script pair into pairs that make sense based
+        on the dominant script, and yield each combination with its dominant script."""
+
+        # First, partition the pair by their assigned scripts
+        allFirstScripts: dict[tuple[str, ...], list[str]] = {}
+        allSecondScripts: dict[tuple[str, ...], list[str]] = {}
+        for glyph in self.firstGlyphs:
+            if glyph not in glyphScripts:
+                glyphScripts[glyph] = {COMMON_SCRIPT}
+            allFirstScripts.setdefault(tuple(glyphScripts[glyph]), []).append(glyph)
+        for glyph in self.secondGlyphs:
+            if glyph not in glyphScripts:
+                glyphScripts[glyph] = {COMMON_SCRIPT}
```

Comment:
@simoncozens any reason the code adds a stand-in script here (modifying a passed variable) instead of upfront?